src/Microsoft.Data.SqlClient/tests/FunctionalTests/SslOverTdsStreamTest.cs

```diff
@@ -0,0 +1,347 @@
+            while (readCount > 0 && offset < bytes.Length);
+            Array.Copy(bytes, 0, output, 0, offset);
+            return offset;
+        }
+
+        private static async Task<int> ReadPacket(Stream tdsStream, Func<Stream, byte[], int, int, Task<int>> action, byte[] output)
+        {
+            int readCount;
+            int offset = 0;
+            byte[] bytes = new byte[TdsEnums.DEFAULT_LOGIN_PACKET_SIZE];
+            do
+            {
+                readCount = await action(tdsStream, bytes, offset, bytes.Length - offset);
+                if (readCount > 0)
+                {
+                    offset += readCount;
+                }
+            }
+            while (readCount > 0 && offset < bytes.Length);
+            Array.Copy(bytes, 0, output, 0, offset);
+            return offset;
+        }
+
+        private static byte[] WritePackets(int encapsulatedPacketCount, int passthroughPacketCount, Action<Stream, int> action)
+        {
+            byte[] buffer = null;
+            using (LimitedMemoryStream stream = new LimitedMemoryStream())
+            {
+                //using (SslOverTdsStream tdsStream = new SslOverTdsStream(stream))
+                using (Stream tdsStream = CreateSslOverTdsStream(stream))
+                {
+                    for (int index = 0; index < encapsulatedPacketCount; index++)
+                    {
+                        action(tdsStream, index);
+                    }
+                    InvokeFinishHandshake(tdsStream);//tdsStream.FinishHandshake();
+                    for (int index = 0; index < passthroughPacketCount; index++)
+                    {
+                        action(tdsStream, encapsulatedPacketCount + index);
+                    }
+                }
+                buffer = stream.ToArray();
+            }
+            return buffer;
+        }
+
+        private static void SetupArrays(int packetCount, out byte[] input, out byte[] output)
+        {
+            byte[] pattern = new byte[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13 };
+            input = new byte[packetCount * TdsEnums.DEFAULT_LOGIN_PACKET_SIZE];
+            output = new byte[input.Length];
+            for (int index = 0; index < packetCount; index++)
+            {
+                int position = 0;
+                while (position < TdsEnums.DEFAULT_LOGIN_PACKET_SIZE)
+                {
+                    int copyCount = Math.Min(pattern.Length, TdsEnums.DEFAULT_LOGIN_PACKET_SIZE - position);
+                    Array.Copy(pattern, 0, input, (TdsEnums.DEFAULT_LOGIN_PACKET_SIZE * index) + position, copyCount);
+                    position += copyCount;
+                }
+            }
+        }
+
+        private static void Validate(byte[] input, byte[] output)
+        {
+            Assert.True(input.AsSpan().SequenceEqual(output.AsSpan()));
+        }
+
+        internal static class TdsEnums
+        {
+            public const int DEFAULT_LOGIN_PACKET_SIZE = 4096;
```

Comment:
Adding the comment from TdsEnum could make it easier to follow.
```
// Login packet can be no greater than 4k until server sends us env-change
        // increasing packet size.
```
Question: what happens when server sends the env-change and packet size is changed? the max packet size is 32768. willt he test fail in that situation?

Reply:
I haven't tested it but I believe that it would. In practice I've only ever seen (and based on the behaviour I expect there to be) only one pre-handshake packet. The test uses more than one packet so that if it ever did manage to happen we have coverage. I don't see a way to change the packetsize mid login.